File: Hacktoberfest2021-main/Data Structure Algo/C/QuickSort.c
#include<stdio.h>
#include<stdlib.h>

void swap(int a[], int b, int c)
{
	int temp;
	temp=a[b];
	a[b]=a[c];
	a[c]=temp;
}

void quicksort(int a[], int left, int right)
{
	int i,j,pivot;
	if(left<right)
	{
		pivot=a[left];
		i=left+1; 
		j=right;
		while(1)
		{
			while(i<right && pivot>=a[i])
                i++;
			while(pivot<a[j])
                j--;
			if(i<j)
				swap(a,i,j);
			else
            { 
				swap(a,left,j);
				break;
			}
		}
		quicksort(a,left,j-1);
		quicksort(a,j+1,right);
	}
}

void main()
{
	int i,n;
    printf("Enter the size of array\n");
	scanf("%d",&n);
	int* a =(int*)malloc(n*sizeof(int));
    if(a== NULL)
    {
        printf("Null pointer\n");
        exit(1);
    }
	printf("Enter the array elements(%d)\n",n);
    
	for(i=0;i<n;i++)
		scanf("%d", &a[i]);
    
	quicksort(a,0,n-1);
	printf("The sorted array is :\n");
	for(i=0;i<n;i++)
		printf("%d  ",a[i]);
    
	printf("\n");
}

File: JavaScript/Sorting-Algorithms/QuickSort.js
// Quick sort works by choosing a pivot element, and making sure that all the elements to the left of
// the pivot element is less than the pivot(not necessarily sorted, they just need to be  less than the pivot)
// and that all the elements to the right of the pivot are all greater than it.

// The best and average time complexity is O(nlogn). But the worst case scenario for the time complexity is O(n^2).
// The space complexity is O(log n).

// Contributed by Rajat(@Rajatm544)

function pivot(arr, start = 0, end = arr.length - 1) {
  // choose an arbitraty pivot element
  let pivot = arr[start];
  let swapIdx = start;

  for (let i = start + 1; i <= end; i++) {
    // If current element is lesser than our pivot element,increment the pivot index
    if (arr[i] < pivot) {
      swapIdx++;

      // Swap current element with the element at the new pivot index
      [arr[i], arr[swapIdx]] = [arr[swapIdx], arr[i]];
    }
  }

  // Swap the pivot element with the element at the swapIndex index
  [arr[swapIdx], arr[start]] = [arr[start], arr[swapIdx]];
  return swapIdx; // Return the index of the pivot element after swapping
}

// funtion to recursively obtain various pivot indices and theryby sort the array
function quickSort(arr, left = 0, right = arr.length - 1) {
  // Base case is that the left and right pointers don't overlap, after which we'll be left with an array of 1 item
  if (left < right) {
    let pivotIndex = pivot(arr, left, right);
    // For left subarray, which is everything to the left of the pivot element
    quickSort(arr, left, pivotIndex - 1);
    // For the right sub array, which is everything to the right of the pivot element
    quickSort(arr, pivotIndex + 1, right);
  }
  // Return the array, when it's of length 1 i.e, left === right
  return arr;
}
